File: day9/day9.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func part_one(line []int) int {
	all_zero := true
	for _, v := range line {
		if v != 0 {
			all_zero = false
		}
	}

	if all_zero {
		return 0
	}
	next := []int{}

	for i := 0; i < len(line)-1; i += 1 {
		next = append(next, line[i+1]-line[i])
	}

	return line[len(line)-1] + part_one(next)
}

func main() {
	start := time.Now()
	file, _ := os.ReadFile("./input.txt")
	as_rows := strings.Split(string(file), "\n")
	part_one_ans := 0
	part_two_ans := 0
	for _, line := range as_rows {
		l := strings.Split(line, " ")
		as_int := []int{}
		for _, c := range l {
			ii, _ := strconv.Atoi(c)
			as_int = append(as_int, ii)
		}
		part_one_ans += part_one(as_int)
		// reverse list for part 2
		for i, j := 0, len(as_int)-1; i < j; i, j = i+1, j-1 {
			as_int[i], as_int[j] = as_int[j], as_int[i]
		}

		part_two_ans += part_one(as_int)
	}

	fmt.Printf("Part one: %d\n", part_one_ans)
	fmt.Printf("Part two: %d\n", part_two_ans)
	fmt.Printf("Time elapsed: %d milliseconds\n", time.Now().UnixMilli()-start.UnixMilli())
}

File: day9/day9.py
# D = open("input.txt").read().strip()
D = open("test.txt").read().strip()
s = D.split("\n")

score = 0
for r in s:
    row = [int(n) for n in r.split(" ")]
    end = []
    diff = [12321]
    while diff and not all(a == 0 for a in diff):
        for i in range(len(row)-1):
            diff.append(row[i+1]-row[i])
        print(diff)
        end.append(row[-1]+diff[-1])
        row = diff
    score += sum(end)
    print(score)
print(score)
        
